sdlib/d/gc/extent.d

```diff
@@ -306,57 +306,127 @@ public:
 		return isAppendableSizeClass(sizeClass);
 	}
 
 	@property
-	ref shared(Bitmap!256) freeSpaceFlags() {
-		assert(isSlab(), "freeSpaceFlags accessed on non slab!");
-		assert(supportsFreeSpace, "freeSpaceFlags not supported!");
+	ref shared(Bitmap!256) slabMetaDataFlags() {
+		assert(isSlab(), "slabMetaDataFlags accessed on non slab!");
+		assert(supportsSlabMetaData, "size class not supports slab metadata!");
 
-		return _metadata.slabData.freeSpaceData.freeSpaceFlags;
+		return _metadata.slabData.slabMetaData.slabMetaDataFlags;
+	}
+
+	bool hasSlabMetaData(uint index) {
+		assert(isSlab(), "hasSlabMetaData accessed on non slab!");
+		assert(index < slotCount, "index is out of range!");
+
+		return supportsSlabMetaData && slabMetaDataFlags.valueAtAtomic(index);
 	}
 
 	ushort* freeSpacePtr(uint index) {
 		assert(isSlab(), "freeSpacePtr accessed on non slab!");
-		assert(supportsFreeSpace, "size class not supports freeSpace!");
+		assert(supportsSlabMetaData, "size class not supports slab metadata!");
 
 		return cast(ushort*) (address + (index + 1) * slotSize) - 2;
 	}
 
 	void setFreeSpace(uint index, size_t freeSpace) {
 		assert(isSlab(), "setFreeSpace accessed on non slab!");
-		assert(freeSpace <= slotSize, "freeSpace exceeds alloc size!");
+		assert(freeSpace <= getTotalCapacity(index),
+		       "freeSpace exceeds alloc size!");
 		assert(index < slotCount, "index is out of range!");
-		assert(supportsFreeSpace, "size class not supports freeSpace!");
+		assert(supportsSlabMetaData, "size class not supports slab metadata!");
 
-		if (freeSpace == 0) {
-			freeSpaceFlags.clearBitAtomic(index);
+		// If we have a finalizer, the freeSpace may be 0:
+		if (freeSpace == 0 && !hasFinalizer(index)) {
+			slabMetaDataFlags.clearBitAtomic(index);
 			return;
 		}
 
 		// Encode freespace and write it to the last byte (or two bytes) of alloc.
 		writePackedFreeSpace(freeSpacePtr(index), freeSpace & ushort.max);
-		freeSpaceFlags.setBitAtomic(index);
+		slabMetaDataFlags.setBitAtomic(index);
 	}
 
 	size_t getFreeSpace(uint index) {
 		assert(isSlab(), "getFreeSpace accessed on non slab!");
 		assert(index < slotCount, "index is out of range!");
 
-		if (!supportsFreeSpace || !freeSpaceFlags.valueAtAtomic(index)) {
+		if (!hasSlabMetaData(index)) {
 			return 0;
 		}
 
 		// Decode freespace, found in the final byte (or two bytes) of the alloc:
 		return readPackedFreeSpace(freeSpacePtr(index));
 	}
 
+	size_t getTotalCapacity(uint index) {
+		assert(isSlab(), "getTotalCapacity accessed on non slab!");
+		assert(index < slotCount, "index is out of range!");
+
+		if (!hasFinalizer(index)) {
+			return slotSize;
+		}
+
+		return slotSize - PointerSize;
+	}
+
+	enum FinalizerBit = nativeToBigEndian!ulong(ulong(0x2));
+
+	bool hasFinalizer(uint index) {
+		assert(isSlab(), "hasFinalizer accessed on non slab!");
+		assert(index < slotCount, "index is out of range!");
+
+		return hasSlabMetaData(index)
+			&& (*(finalizerPtr(index)) & FinalizerBit);
```

Comment:
I'd say you want to do the checks int he opposite order, this way you can avoid snooping in the metadata when you don't have to.

Reply:
In fact, this is likely to be false, consider that:
  - If the slot is not full, then this is probably going to be zero.
  - If it has metadata, in most cases, it won't have a finalizer: 0.
  - If there is a pointer in there, it'll the MSB, which are going to be zero.
  - If there is a number there, once again, MSB, and most numbers are small, so it's likely to be zero.
  - If there is something else in there that don't heavily bias toward zero (float, random/compressed data) we are still at 50/50.
  
We should expect to find a zero in there the VAST majority of the time.

Reply:
BTW, this probably warrant a comment :P